### DISCUSSION TITLE
`:setup-children` for nested groups?

### DISCUSSION THREAD

**Discussion by mekeor** [magit/transient#291] [Support] [answered]
This nested, matrix-like, two-dimensional transient works fine:

```elisp
(defun example () (interactive) (message "example"))

(transient-define-prefix example-static ()
  [ [ "Left column"
      ("d" "d" example :transient t)
      ("f" "f" example :transient t)]
    [ "Right column"
      ("j" "j" example :transient t)
      ("k" "k" example :transient t)]])
```

But I want to determine the groups dynamically, when the transient is called. I tried this:

```elisp
(transient-define-prefix example-dynamic ()
  [ :setup-children
    (lambda (_)
      (transient-parse-suffixes
       'not-sure-about-this
       [ :class transient-row
         [ "Left column"
           ("d" "d" example :transient t)
           ("f" "f" example :transient t)]
         [ "Right column"
           ("j" "j" example :transient t)
           ("k" "k" example :transient t)]]))])
```

But it calling this transient results in this error:

```
transient-setup: Lisp nesting exceeds `max-lisp-eval-depth': 1601
```

I also tried other variations. For example, when I remove `:class transient-row`, I get this error:

```
transient-setup: No applicable method: transient-format, #s(transient-column 1 nil nil nil nil nil nil nil nil ...)
```

What am I doing wrong? How can I dynamically build a two-dimensional transient?

**Comment by tarsius** [comment, marked as answer]
```elisp
(transient-define-prefix example-dynamic ()
  [:class transient-columns
   :setup-children
   (lambda (_)
     (transient-parse-suffixes
      'example-dynamic
      '(["Left column"
         ("d" "d" transient-echo-arguments :transient t)
         ("f" "f" transient-echo-arguments :transient t)]
        ["Right column"
         ("j" "j" transient-echo-arguments :transient t)
         ("k" "k" transient-echo-arguments :transient t)])))])
```

> This nested, matrix-like, two-dimensional transient

Such layouts have to use the `transient-columns` group type, there is no other group type that supports more than one dimension.

When using `:setup-children` the class has to be specified but you picked the wrong class and specified it in the wrong place.

Using `[["Left column" ...` would actually work, but using `'(["Left column" ...` is better because it discourages the mentioned mistakes.

I'll have to think about how to make it more obvious in the manual. I might also add an example.

**Reply by tarsius** [reply]
I've added the example in cfd17a714b0ef9401d570946795a38fe1d8d6e08.
